db/db_impl.cc

```diff
@@ -242,6 +243,49 @@ DBImpl::DBImpl(const DBOptions& options, const std::string& dbname,
   // we won't drop any deletion markers until SetPreserveDeletesSequenceNumber()
   // is called by client and this seqnum is advanced.
   preserve_deletes_seqnum_.store(0);
+
+  error_handler_.reset(new ErrorHandler(immutable_db_options_, &mutex_));
+}
+
+Status DBImpl::Resume() {
+  InstrumentedMutexLock db_mutex(&mutex_);
+
+  if (!error_handler_->IsDBStopped() && !error_handler_->IsBGWorkStopped()) {
+    // Nothing to do
+    return Status::OK();
+  }
+
+  Status s = error_handler_->GetBGError();
+  if (s.severity() > Status::Severity::kHardError) {
+    ROCKS_LOG_INFO(immutable_db_options_.info_log,
+        "DB resume requested but failed due to Fatal/Unrecoverable error");
+    return s;
+  }
+
+  ROCKS_LOG_INFO(immutable_db_options_.info_log,
+                 "Resuming DB");
+
+  JobContext job_context(0);
+  FindObsoleteFiles(&job_context, true);
+  error_handler_->ClearBGError();
+  mutex_.Unlock();
+
+  job_context.manifest_file_number = 1;
+  if (job_context.HaveSomethingToDelete()) {
+    PurgeObsoleteFiles(job_context);
+  }
+  job_context.Clean();
+
+  mutex_.Lock();
+  if (bg_work_paused_ == 0) {
+    MaybeScheduleFlushOrCompaction();
+  }
+
+  // No need to check BGError again. If something happened, event listener would be
+  // notified and the operation causing it would have failed
+  ROCKS_LOG_INFO(immutable_db_options_.info_log,
+      "Successfully resumed DB");
```

Comment:
We usually try to avoid logging inside DB mutex. This applies to multiple logging in this function. We may not care that in this case where DB is already stopped, but it will be great that we always keep this practice when possible.